src/connection.cpp

```diff
@@ -740,14 +865,41 @@ namespace oxen::quic
         return 0;
     }
 
+    // this callback is defined for debugging datagrams
+    int Connection::ack_datagram(uint64_t dgram_id)
+    {
+        log::trace(log_cat, "Connection (CID: {}) acked datagram ID:{}", _source_cid, dgram_id);
+        return 0;
+    }
+
     int Connection::recv_datagram(bstring_view data, bool fin)
     {
         log::trace(log_cat, "Connection (CID: {}) received datagram: {}", _source_cid, buffer_printer{data});
 
-        // append received datagram to tetris_buffer if packet_splitting is enabled
-        if (_packet_splitting)
+        std::optional<bstring> maybe_data;
+        uint16_t dgid = oxenc::load_big_to_host<uint16_t>(data.data());
+
+        if (dgid % 4 == 0)
+            log::trace(log_cat, "Datagram sent unsplit, bypassing rotating buffer");
+        else if (_packet_splitting)
         {
-            // TODO: should this logic go earlier in the process inside handle_packet?
+            try
+            {
+                // send received datagram to tetris_buffer if packet_splitting is enabled
+                maybe_data = datagrams->to_buffer(data, dgid);
+            }
+            catch (std::exception& e)
+            {
+                log::error(log_cat, "Exception: {}", e.what());
+                return -1;
```

Comment:
Should this be NGTCP2_ERR_CALLBACK_FAILURE?

Reply:
that's a good question... on one hand it doesn't matter because any non-zero return value buffs out to `NGTCP2_ERR_CALLBACK_FAILURE` [upon return](https://github.com/ngtcp2/ngtcp2/blob/d80f9b3d709da79cdbd2b94acfae3073810de80d/lib/ngtcp2_conn.c#L589-L593) so we might as well return that... on the other hand that specific line isn't a failure in the callback, but in the buffer.

maybe we could extend std exception and make our own that we throw and handle internally?